### PR TITLE
feature(pu): add il and rl demo for gfootball

### PR DESCRIPTION
**Pull request by puyuan1996**
## Description
- adapt gfootball env to DI-engine env interface.
- add gfootball imitation learning config, use the rule_based_bot and kaggle_5th_place model to collect demo data.
- add episode_to_transitions_filter to only use the demo data whose return is larger than a given threshold in il demo.
- add dqn config for gfootball.

## Related Issue

https://github.com/opendilab/DI-engine/issues/335

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: dizoo/gfootball/entry/gfootball_dqn_config.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        discount_factor=0.997,
+        model=dict(),
+        learn=dict(
+            update_per_collect=20,
+            batch_size=512,
+            learning_rate=0.0001,
+            target_update_freq=500,
+        ),
+        collect=dict(n_sample=256),
+        eval=dict(evaluator=dict(eval_freq=5000, n_episode=evaluator_env_num)),
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/gfootball_dqn_config.py:30]: why `n_episode` here

File: dizoo/gfootball/entry/gfootball_il_config.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ),
+    policy=dict(
+        test_accuracy=False,
+        # Note, only if test_accuracy=True, we will test accuracy in train dataset and validation dataset
+        # use the il model in the path <il_model_path>.
+        # Users should add their own il model path here. Model path should lead to a model.
+        # Absolute path is recommended. In DI-engine, it is ``exp_name/ckpt/ckpt_best.pth.tar``.
+        il_model_path='il_model_path_placeholder',
+        cuda=True,
+        nstep=1,
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/gfootball_il_config.py:25]: remove unused key like `nstep`

File: dizoo/gfootball/entry/gfootball_il_config.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        discount_factor=0.97,
+        model=dict(),
+        learn=dict(
+            update_per_collect=20,
+            batch_size=512,
+            learning_rate=0.0001,
+            target_update_freq=500,
+            learner=dict(load_path=None),
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/gfootball_il_config.py:33]: remove this

File: dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+"""
+phase 1: train/obtain expert policy
+"""
+train_config = [deepcopy(gfootball_il_main_config), deepcopy(gfootball_il_create_config)]
+input_cfg = train_config
+if isinstance(input_cfg, str):
+    cfg, create_cfg = read_config(input_cfg)
+else:
+    cfg, create_cfg = input_cfg
+create_cfg.policy.type = create_cfg.policy.type + '_command'
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py:30]: remove codes not used here, such as `command`

File: dizoo/gfootball/model/bots/TamakEriFever/football_ikki.py [hunk 5]
```diff
@@ -248,8 +247,9 @@ def forward(self, x):
 
             p = F.relu(torch.cat([p_self, p_opp], dim=1))
             s_concat = s.view(bs, 1, -1).repeat(1, p.size(1), 1)
-            p = torch.cat([p, x['distance']['p2bo'].view(bs, p.size(1), -1), s_concat], dim=2)
-
+            # p = torch.cat([p, x['distance']['p2bo'].view(bs, p.size(1), -1), s_concat], dim=2)
+            # TODO(pu)
+            p = torch.cat([p, x['distance']['p2bo'].repeat(1,2,1).view(bs, p.size(1), -1), s_concat], dim=2)
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/bots/TamakEriFever/football_ikki.py:252]: correct format

File: dizoo/gfootball/model/q_network/football_q_network_default_config.yaml [hunk 6]
```diff
@@ -54,6 +54,7 @@ model:
                 output_dim: 1
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/football_q_network_default_config.yaml:54]: if yaml is not used, remove it

File: dizoo/gfootball/model/q_network/football_q_network.py [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+from typing import Union, Optional, Dict, Callable, List
+import torch
+import torch.nn as nn
+
+from ding.torch_utils import get_lstm, one_hot, to_tensor, to_ndarray
+from ding.utils import MODEL_REGISTRY, SequenceType, squeeze
+from ding.model.common import FCEncoder, ConvEncoder, DiscreteHead, DuelingHead, MultiHead
+from ding.model.template.q_learning import parallel_wrapper
+
+with open(os.path.join(os.path.dirname(__file__), 'football_q_network_default_config.yaml')) as f:
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/football_q_network.py:18]: why not use python config, which has been already implemented

File: dizoo/gfootball/model/q_network/football_q_network.py [hunk 8]
```diff
@@ -0,0 +1,779 @@
...
+            player_encodings = self.player_encoder(x['players'])
+        encoding_list = list(scalar_encodings.values()) + [player_encodings]
+        x = torch.cat(encoding_list, dim=1)
+        # football obs encoding: shape (1, 1312)
+        return x
+
+    def forward(
+            self, inputs: Dict, inference: bool = False, saved_state_timesteps: Optional[list] = None
+    ) -> Dict:
+        r"""
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/football_q_network.py:197]: remove DRQN related comment if you have updated them for football

File: dizoo/gfootball/model/q_network/football_q_network.py [hunk 9]
```diff
@@ -0,0 +1,779 @@
...
+            # This key is used in qtran, the algorithm requires to retain all h_{t} during training
+            x['hidden_state'] = torch.cat(hidden_state_list, dim=-3)
+            if saved_state_timesteps is not None:
+                # the selected saved hidden states, including the hidden state (h) and the cell state (c)
+                x['saved_state'] = saved_state
+            return x
+
+
+@MODEL_REGISTRY.register('football_ngu')
+class FootballNGU(nn.Module):
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/football_q_network.py:302]: remove ngu code in remote repo

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/football_bc.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
+from typing import Tuple, List, Dict, Any
```
Comment (PaParaZz1) [inline ding/policy/football_bc.py:1]: move football bc to dizoo

File: ding/entry/serial_entry_bc.py [hunk 11]
```diff
@@ -96,6 +96,15 @@ def serial_pipeline_bc(
         if stop:
             break
 
+    if not cont and cfg.policy.learn.show_accuracy:
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/entry/serial_entry_bc.py:99]: 这里是为了训练完成后，测试训练好的模型在整个数据集上的准确率，由于是将总数据集分成很多个小的minibatch进行测试，然后求平均得到整个训练集的准确率，好像只能写在训练循环外面？

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/gfootball/entry/gfootball_il_config.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+gfootball_il_create_config = dict(
+    env=dict(
+        type='gfootball',
+        import_names=['dizoo.gfootball.envs.gfootball_env'],
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='bc'),
+)
+gfootball_il_create_config = EasyDict(gfootball_il_create_config)
+create_config = gfootball_il_create_config
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply dizoo/gfootball/entry/gfootball_il_config.py:65]: 由于有收集样本，训练，测试的3个阶段，是写在了同目录下的 x_main.py里面

File: dizoo/gfootball/entry/gfootball_dqn_config.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+)
+gfootball_dqn_main_config = EasyDict(gfootball_dqn_main_config)
+main_config = gfootball_dqn_main_config
+
+gfootball_dqn_create_config = dict(
+    env=dict(
+        type='gfootball',
+        import_names=['dizoo.gfootball.envs.gfootball_env'],
+    ),
+    env_manager=dict(type='base'),
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply dizoo/gfootball/entry/gfootball_dqn_config.py:50]: gfootball obs 包括 {processed_obs'':X,  'raw_obs',X}结构比较复杂，目前还是没有适配好subprocess, 之后我再试试

File: ding/policy/bc.py [hunk 14]
```diff
@@ -58,7 +71,15 @@ def _init_learn(self):
...
-            self._loss = nn.CrossEntropyLoss()
+            if self._cfg.learn.ce_label_smooth:
+                self._loss = LabelSmoothCELoss(0.1)
+            else:
+                self._loss = nn.CrossEntropyLoss()
+
+            if self._cfg.learn.show_accuracy:
+                # accuracy statistics for debugging in discrete action space env, e.g. for gfootball
+                self.total_accuracy_in_dataset = []
+                self.action_accuracy_in_dataset = {k: [] for k in range(19)}
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/policy/bc.py]: ok, changed

File: dizoo/gfootball/entry/gfootball_il_config.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply dizoo/gfootball/entry/gfootball_il_config.py:65]: add some comments to give guidance

File: dizoo/gfootball/entry/gfootball_dqn_config.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply dizoo/gfootball/entry/gfootball_dqn_config.py:50]: try subprocess with `shared_memory=False`?

File: ding/policy/bc.py [hunk 15]
```diff
@@ -31,25 +34,36 @@ def default_model(self) -> Tuple[str, List[str]]:
...
         eval=dict(),
         other=dict(replay_buffer=dict(replay_buffer_size=10000, )),
     )
 
     def _init_learn(self):
-        self._optimizer = Adam(
-            self._model.parameters(),
-            lr=self._cfg.learn.learning_rate,
-        )
+        if not hasattr(self._cfg.learn, 'weight_decay') or self._cfg.learn.weight_decay is None:
```
Comment (PaParaZz1) [inline ding/policy/bc.py:53]: don't use hasttr, you have added it into default_config

File: dizoo/gfootball/model/q_network/tests/test_config.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import pytest
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/tests/test_config.py:1]: remove this

File: dizoo/gfootball/model/q_network/tests/test_football_model.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+import torch
+import os
+import yaml
+from easydict import EasyDict
+from dizoo.gfootball.model.q_network.football_q_network import FootballNaiveQ
+from ding.torch_utils import to_tensor, to_dtype
+from dizoo.gfootball.envs.fake_dataset import FakeGfootballDataset
+import pprint
+
+with open(os.path.join(os.path.abspath(os.path.dirname(os.path.dirname(__file__))),
```
Comment (PaParaZz1) [inline dizoo/gfootball/model/q_network/tests/test_football_model.py:12]: move to python config

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/gfootball/entry/gfootball_dqn_config.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (puyuan1996) [reply dizoo/gfootball/entry/gfootball_dqn_config.py]: ok, thanks

File: dizoo/gfootball/entry/gfootball_il_config.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (puyuan1996) [reply dizoo/gfootball/entry/gfootball_il_config.py:66]: ok

File: ding/policy/bc.py (hunk 15 above)
Comment (puyuan1996) [reply ding/policy/bc.py:53]: ok, thanks

File: dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
+from copy import deepcopy
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py:1]: rename il to bc

File: ding/policy/bc.py (hunk 15 above)
Comment (puyuan1996) [reply ding/policy/bc.py:53]: 在https://github.com/opendilab/DI-engine/blob/1fae1d177fe6f87e104d39df69003344f8588fe2/ding/entry/tests/test_serial_entry_bc.py#L137 这里如果去掉hasattr(self._cfg.learn, 'weight_decay')会出错，好像是dqn_bc这样的策略不能继承bc默认的config?

File: dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py (hunk 18 above)
Comment (puyuan1996) [reply dizoo/gfootball/entry/gfootball_il_kaggle5th_main.py:1]: ok

File: ding/policy/bc.py (hunk 15 above)
Comment (PaParaZz1) [reply ding/policy/bc.py:53]: modify dqn_bc like [this](https://github.com/opendilab/DI-engine/blob/main/ding/world_model/base_world_model.py#L63)

File: dizoo/gfootball/entry/show_dataset.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+dir_path = os.path.dirname(path)
+
+if __name__ == "__main__":
+    config = [main_config, create_config]
+    input_cfg = config
+    if isinstance(input_cfg, str):
+        cfg, create_cfg = read_config(input_cfg)
+    else:
+        cfg, create_cfg = input_cfg
+    create_cfg.policy.type = create_cfg.policy.type + '_command'
```
Comment (PaParaZz1) [inline dizoo/gfootball/entry/show_dataset.py:18]: command

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**